lib/stack_master/role_assumer.rb

```diff
@@ -0,0 +1,59 @@
+require 'active_support/core_ext/object/deep_dup'
+
+module StackMaster
+  class RoleAssumer
+    BlockNotSpecified = Class.new(StandardError)
+
+    def initialize
+      @credentials = {}
+    end
+
+    def assume_role(account, role, &block)
+      raise BlockNotSpecified unless block_given?
+      raise ArgumentError, "Both 'account' and 'role' are required to assume a role" if account.nil? || role.nil?
+
+      original_aws_config = replace_aws_global_config
+      Aws.config[:credentials] = assume_role_credentials(account, role)
+      begin
+        original_cf_driver = replace_cf_driver
+        block.call
+      ensure
+        restore_aws_global_config(original_aws_config)
+        restore_cf_driver(original_cf_driver)
+      end
+    end
+
+    private
+
+    def replace_aws_global_config
```

Comment:
I think the replace/restore functions are not very pretty. Part of that logic also exists in `assume_role`.

How about:

```ruby
def assume_role(...)
...
  role_credentials = assume_role_credentials(account, role)
  with_temporary_role(role_credentials) do
    with_temporary_driver do
      block.call
    end
  end
...
end

def with_temporary_role(credentials, &block)
  original_aws_config = Aws.config
  Aws.config = original_aws_config.deep_dup
  block.call
ensure
  Aws.config = original_aws_config
end

def with_temporary_driver(&block)
  original_driver = StackMaster.cloud_formation_driver
  StackMaster.cloud_formation_driver = original_driver.class.new
  block.call
ensure
  StackMaster.cloud_formation_driver = original_driver
end
```

Reply:
Great idea. Done in 95a8825.